packages/effect/src/internal/stream.ts

```diff
@@ -5678,6 +5678,79 @@ export const scoped = <A, E, R>(
 ): Stream.Stream<A, E, Exclude<R, Scope.Scope>> =>
   new StreamImpl(channel.ensuring(channel.scoped(pipe(effect, Effect.map(Chunk.of))), Effect.void))
 
+/** @internal */
+export const share = dual<
+  <A, E>(config: {
+    readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
+  }) => <R>(
+    self: Stream.Stream<A, E, R>
+  ) => Effect.Effect<Stream.Stream<A, E>, never, R | Scope.Scope>,
+  <A, E, R>(
+    self: Stream.Stream<A, E, R>,
+    config: {
+      readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
+    }
+  ) => Effect.Effect<Stream.Stream<A, E>, never, R | Scope.Scope>
+>(
+  2,
+  <A, E, R>(
+    self: Stream.Stream<A, E, R>,
+    config: {
+      readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
+    }
+  ) =>
+    Effect.gen(function*() {
+      const connector = yield* (config.connector ?? PubSub.unbounded<Take.Take<A, E>>())
+      yield* Effect.forkScoped(runIntoPubSub(self, connector))
+      return flattenTake(fromPubSub(connector))
+    })
+)
+
+/** @internal */
+export const shareRefCount = dual<
+  <A, E>(config: {
+    readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
+  }) => <R>(
+    self: Stream.Stream<A, E, R>
+  ) => Stream.Stream<A, E, R>,
+  <A, E, R>(
+    self: Stream.Stream<A, E, R>,
+    config: {
+      readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
+    }
+  ) => Stream.Stream<A, E, R>
+>(
+  2,
+  <A, E, R>(
+    self: Stream.Stream<A, E, R>,
+    config: {
+      readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
+    }
+  ): Stream.Stream<A, E, R> => {
+    let consumersCount = 0
+    let connector: PubSub.PubSub<Take.Take<A, E>> | null = null
+    let fiber: Fiber.RuntimeFiber<void> | null = null
+    return Effect.gen(function*() {
+      consumersCount++
+      connector ??= yield* (config.connector ?? PubSub.unbounded<Take.Take<A, E>>())
```

Comment:
It should probably default to `PubSub.bounded(16)`, so back-pressure is applied.